stackslib/src/chainstate/stacks/boot/pox_4_tests.rs

```diff
@@ -5076,165 +5083,1872 @@ fn delegate_stack_increase() {
     assert_eq!(&reward_entry.signer.unwrap(), signer_pk_bytes.as_slice());
 }
 
-pub fn get_stacking_state_pox_4(
-    peer: &mut TestPeer,
-    tip: &StacksBlockId,
-    account: &PrincipalData,
-) -> Option<Value> {
-    with_clarity_db_ro(peer, tip, |db| {
-        let lookup_tuple = Value::Tuple(
-            TupleData::from_data(vec![("stacker".into(), account.clone().into())]).unwrap(),
-        );
-        let epoch = db.get_clarity_epoch_version().unwrap();
-        db.fetch_entry_unknown_descriptor(
-            &boot_code_id(boot::POX_4_NAME, false),
-            "stacking-state",
-            &lookup_tuple,
-            &epoch,
-        )
+// In this test two solo stacker-signers Alice & Bob sign & stack
+//  for two reward cycles. Alice provides a signature, Bob uses
+//  'set-signer-key-authorizations' to authorize. Two cycles later,
+//  when no longer stacked, they both try replaying their auths.
+#[test]
+fn test_scenario_one() {
+    // Alice solo stacker-signer setup
+    let mut alice = StackerSignerInfo::new();
+    // Bob solo stacker-signer setup
+    let mut bob = StackerSignerInfo::new();
+
+    let default_initial_balances = 1_000_000_000_000_000_000;
+    let observer = TestEventObserver::new();
+    let test_signers = TestSigners::default();
+    let mut initial_balances = vec![
+        (alice.principal.clone(), default_initial_balances),
+        (bob.principal.clone(), default_initial_balances),
+    ];
+    let aggregate_public_key = test_signers.aggregate_public_key.clone();
+    let mut peer_config = TestPeerConfig::new(function_name!(), 0, 0);
+    let private_key = peer_config.private_key.clone();
+    let addr = StacksAddress::from_public_keys(
+        C32_ADDRESS_VERSION_TESTNET_SINGLESIG,
+        &AddressHashMode::SerializeP2PKH,
+        1,
+        &vec![StacksPublicKey::from_private(&private_key)],
+    )
+    .unwrap();
+
+    // reward cycles are 5 blocks long
+    // first 25 blocks are boot-up
+    // reward cycle 6 instantiates pox-3
+    // we stack in reward cycle 7 so pox-3 is evaluated to find reward set participation
+    peer_config.aggregate_public_key = Some(aggregate_public_key.clone());
```

Comment:
I think a lot of the logic around the scenario tests have the same boot up logic around setting up test signers, and the peer config. Maybe we can pull that all out into a llhelper start up function. I know we have a ton of these already throughout the code, but maybe call it "pox_4_scenario_test_peer()" and then make sure its got some clear documentation about what it does and why its different from the other functions. Main thing I think is slightly longer reward cycle lengths and prepare lengths and then that it sets a very distance epoch 3 activation as none of these pox 4 tests need to advance into epoch 3 to work.

Reply:
Copy, refactored the setup out into 'pox_4_scenario_test_setup'.